### PR TITLE
[Bug] Ignore zero-query padding rows in Mamba prefill metadata

### PR DESCRIPTION
**Pull request by izikgo**
## Purpose

This PR prevents full-CUDA-graph padding rows from being classified as Mamba
prefill rows.

Full CUDA graph decode can pad a decode batch up to a captured graph size. The
padded rows are not scheduled requests and have zero query tokens. Today, when
Mamba calls `split_decodes_and_prefills(..., treat_short_extends_as_decodes=False)`,
the splitter ORs `is_prefilling` directly into the prefill mask. If a padded
zero-query row carries stale prefill counters, it can become a fake Mamba
prefill.

That fake prefill matters because Mamba's full-CUDA-graph decode path keeps a
persistent decode `state_indices_tensor_d` buffer and refreshes it only for pure
decode batches. Once the padding row makes `num_prefills > 0`, the refresh is
skipped. CUDA graph replay can then use stale nonzero Mamba state slots for live
decode rows after the request-to-row mapping changes, causing a live request to
read/write another request's recurrent state.

The fix has two parts:

1. In `split_decodes_and_prefills()`, only allow `is_prefilling` to mark real
   rows with `query_lens > 0` as prefills.
2. In `GPUModelRunner`, explicitly clear `is_prefilling` for padded rows beyond
   `num_reqs`.

The consumer-side guard is the correctness fix; the producer-side clearing makes
the common metadata invariant explicit for padded rows.

This PR was prepared with AI assistance and manually reviewed.

## Test Plan

Added a focused splitter regression test:

```bash
python3 -m pytest tests/v1/attention/test_attention_splitting.py \
  -k stale_padded_is_prefilling -q
```

The test builds a padded decode batch with positive-query decode rows and
zero-query padding rows, then marks a padded row as `is_prefilling=True`.
Expected result: the padded row is ignored as prefill work, so the split remains
decode-only.

I also ran a minimal in-process `LLM` reproducer (`llm_mamba_cg_repro.py`) that
loads a small Mamba model, forces a tiny full-CUDA-graph batch size, creates a
chunked-prefill padding row, then aborts one decode request to change the live
row-to-state mapping. Its detector wraps the Mamba CUDA-graph metadata update
and reports when a fake padded prefill causes a stale persistent state-index
buffer to be replayed instead of refreshed.

Minimal in-process LLM reproducer:

```python
#!/usr/bin/env python3
"""Minimal in-process LLM reproducer for the Mamba full-CUDA-graph state bug."""

from __future__ import annotations

import os

os.environ.setdefault("VLLM_USE_V1", "1")
os.environ.setdefault("VLLM_ENABLE_V1_MULTIPROCESSING", "0")

import torch

from vllm import LLM, SamplingParams
from vllm.sampling_params import RequestOutputKind
from vllm.v1.attention.backends import mamba_attn
from vllm.v1.attention.backends.utils import NULL_BLOCK_ID

phase = "init"
saw_bug = False

MODEL = os.getenv("MODEL", "tiiuae/Falcon-H1-0.5B-Base")
MAMBA_BACKEND = os.getenv("MAMBA_BACKEND", "flashinfer")
ATTENTION_BACKEND = os.getenv("ATTENTION_BACKEND", "FLASHINFER")


def cpu(t: torch.Tensor | None) -> torch.Tensor | None:
    return None if t is None else t.detach().cpu().clone()


def head(t: torch.Tensor | None) -> list | None:
    return None if t is None else t[:4].tolist()


def install_detector() -> None:
    original = (
        mamba_attn.BaseMambaAttentionMetadataBuilder
        ._update_metadata_for_cudagraph_capture
    )

    def wrapped(self, metadata):  # type: ignore[no-untyped-def]
        global saw_bug

        incoming = cpu(metadata.state_indices_tensor_d)
        prefill = cpu(metadata.state_indices_tensor_p)
        before = cpu(self.state_indices_tensor_d[:metadata.num_reqs])

        n = min(metadata.num_decodes, incoming.shape[0], before.shape[0])
        w = min(incoming.shape[1], before.shape[1]) if n else 0
        stale = w > 0 and not torch.equal(before[:n, :w], incoming[:n, :w])

        result = original(self, metadata)

        after = cpu(self.state_indices_tensor_d[:metadata.num_reqs])
        refreshed = stale and w > 0 and torch.equal(after[:n, :w], incoming[:n, :w])
        fake_prefill = (
            metadata.num_decodes > 0
            and metadata.num_prefills > 0
            and prefill is not None
            and prefill.numel() > 0
            and bool(torch.all(prefill == NULL_BLOCK_ID).item())
        )

        if phase == "after_decode_abort" and fake_prefill and stale and not refreshed:
            saw_bug = True

        if phase == "after_decode_abort" and (fake_prefill or stale or refreshed):
            print(
                "[detector]",
                f"fake_prefill={fake_prefill}",
                f"stale={stale}",
                f"refreshed={refreshed}",
                f"incoming={head(incoming)}",
                f"before={head(before)}",
                f"after={head(after)}",
                flush=True,
            )
        return result

    (
        mamba_attn.BaseMambaAttentionMetadataBuilder
        ._update_metadata_for_cudagraph_capture
    ) = wrapped


def sampling(max_tokens: int) -> SamplingParams:
    return SamplingParams(
        temperature=0.0,
        max_tokens=max_tokens,
        min_tokens=max_tokens,
        ignore_eos=True,
        output_kind=RequestOutputKind.FINAL_ONLY,
    )


def long_token_prompt(llm: LLM, tokens: int) -> dict[str, list[int]]:
    text = "chunked prefill filler alpha beta gamma delta epsilon. "
    tokenizer = llm.get_tokenizer()
    ids: list[int] = []
    while len(ids) < tokens:
        ids.extend(tokenizer.encode(text, add_special_tokens=False))
    return {"prompt_token_ids": ids[:tokens]}


def step(llm: LLM, count: int) -> None:
    for _ in range(count):
        llm.llm_engine.step()
        if saw_bug:
            return


def main() -> int:
    global phase

    install_detector()
    llm = LLM(
        model=MODEL,
        trust_remote_code=True,
        dtype="bfloat16",
        max_model_len=1024,
        max_num_seqs=4,
        max_num_batched_tokens=64,
        enable_chunked_prefill=True,
        max_cudagraph_capture_size=4,
        cudagraph_capture_sizes=[4],
        compilation_config={"cudagraph_mode": "FULL"},
        attention_backend=ATTENTION_BACKEND,
        mamba_backend=MAMBA_BACKEND,
        mamba_cache_mode="none",
        mamba_ssm_cache_dtype="float32",
        enforce_eager=False,
        disable_log_stats=True,
    )

    phase = "decode_warmup"
    decode_ids = llm.enqueue(
        [f"request {i}: 1 2 3 4 5" for i in range(3)],
        [sampling(512)] * 3,
        use_tqdm=False,
    )
    step(llm, 6)

    phase = "prefill"
    prefill_id = llm.enqueue(
        [long_token_prompt(llm, 768)], [sampling(16)], use_tqdm=False
    )[0]
    step(llm, 1)

    phase = "after_prefill_abort"
    llm.llm_engine.abort_request([prefill_id], internal=True)
    step(llm, 5)

    phase = "after_decode_abort"
    llm.llm_engine.abort_request([decode_ids[0]], internal=True)
    step(llm, 12)
    llm.llm_engine.abort_request(decode_ids[1:], internal=True)

    if saw_bug:
        print("[result] BUG REPRODUCED: stale fake prefill skipped state refresh")
        return 0
    print("[result] NO SIGNAL")
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
```

## Test Result

Pytest:

```bash
python3 -m pytest tests/v1/attention/test_attention_splitting.py \
  -k stale_padded_is_prefilling -q
```

Supplemental minimal `LLM` reproducer:

```text
[detector] fake_prefill=true stale_skip=true refresh_changed=false
           after_matches_incoming=false
[result] BUG REPRODUCED: fake prefill skipped CG state refresh
```

### CODE REVIEW

**Comment by izikgo** [comment]
Closing as superseded by #41873, which merged the producer-side fix for stale is_prefilling in padded CUDA graph rows.
